### DISCUSSION TITLE
[SYCL][CUDA] assert when buffer size is 0 ?

### DISCUSSION THREAD

**Discussion by zjin-lcf** [intel/llvm#5156] [General]
When the size of a device buffer is zero, the runtime will show assertion error.  Could the assertion occur only when the size is negative ?

```
  what():  SYCL buffer size is zero. To create a device accessor, SYCL buffer size must be greater than zero. -30 
```

In CUDA, cudaMalloc() does not return an error when the size is zero. 

Porting a CUDA program to SYCL using SYCL buffers is smooth when a buffer size of zero does not cause the assert error.

**Comment by AerialMantis** [comment]
So zero-sized buffers result in an error as it was previously the case in SYCL 2020 that zero-sized buffers were not permitted, even if the backend could allow it.

However, the SYCL 2020 spec was recently amended to permit zero-sized buffers:

```
Zero-sized buffers are permitted. In this case, the buffer still handles ownership as normal, but does not
need to store the zero-sized data itself.
```

As this is now permitted by the spec I think it's a valid enhancement, and I believe this is a general issue which would apply to all backends, so not just for the CUDA backend.

cc @bader

**Reply by bader** [reply]
This seemed to be working 2.5 years ago - https://github.com/intel/llvm/issues/123.

**Reply by bader** [reply]
Is the problem with creating an accessor to zero-size buffer?

>  what():  SYCL buffer size is zero. To create a device accessor, SYCL buffer size must be greater than zero. -30

**Reply by AerialMantis** [reply]
So I think in the past this wasn't well defined and so some implementations allowed it, but then it was clarified as not permitted, and then more recently it was decided that it should be permitted.

Yeah I believe the problem comes from the creation of an accessor with a zero-sized range - https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/accessor.hpp#L1671

**Reply by zjin-lcf** [reply]
I should not call data copy between a host and a device when the size of a buffer is zero.

Thanks.

**Reply by keryell** [reply]
If we allow buffers of size 0, why not allowing an accessor of size 0 we cannot use either?
The copy of 0 element would work since there would be no copy.

**Reply by AerialMantis** [reply]
I believe that is the intention of the recent clarification to the specification, the idea being that creating an accessor to a zero-sized buffer would still represent a data dependency but no copy would be performed as there is no data to copy. I think we should change this so that creating an accessor from a zero-sized buffer is valid to conform with the change to the specification.

**Reply by alexbatashev** [reply]
What's the use case for buffers and accessors anyway? You can't really use them, so, this sounds more like a great opportunity to hurt yourself in a hard-to-catch manner. Right now buffer clearly indicates to users what's wrong, eliminating need for bounds checking all over application.

**Reply by zjin-lcf** [reply]
Please take a look at https://github.com/zjin-lcf/HeCBench/tree/master/mnist-sycl
I am not sure if you think this is a good example to show the use case.   
As a workaround,  copy is disabled when the copy size is 0 in my example.

**Reply by AerialMantis** [reply]
So I believe the use case for this was mainly for writing generic code, where the size may or may not be zero, @alexbatashev you're right this does introduce a requirement on the user to at some point check to avoid dereferencing the accessor as this would be undefined behaviour.

**Reply by zjin-lcf** [reply]
Running this example would also also segfault without the workaround proposed in #4960 
My comment is not relevant to the title of the topic, but I feel people need to be aware of it.  Thanks.
